pkg/thing/interactors/register_thing.go

```diff
@@ -82,24 +83,27 @@ func (i *ThingInteractor) Register(authorization, id, name string) error {
 	if err != nil {
 		errReply := i.reply(id, "", err)
 		if errReply != nil {
-			i.logger.Error(errReply)
+			return fmt.Errorf("error sending success response to client: %v: %w", errReply, err)
 		}
-
-		return err
+		return fmt.Errorf("error registering thing: %w", err)
 	}
 
 	// Get the id generated as a token and send in the response
 	token, err := i.thingProxy.Create(id, name, authorization)
 	errReply := i.reply(id, token, err)
+	if err != nil {
+		if errReply != nil {
+			return fmt.Errorf("error sending success response to client: %v: %w", errReply, err)
+		}
+		return fmt.Errorf("error registering thing: %w", err)
```

Comment:
Please, follow the DRY (Don't Repeat Yourself) rule. Notice what you're doing here is too similar to lines +85 to +88. My suggestions:

- Let's take advantage of the fact the method `reply` is already expecting an error as one of its attributes. i.e. we can append the sending error information there instead of doing this in the `Register` method. For example:

```go
	err := i.verifyThingID(id)
	if err != nil {
		return i.reply(id, "", fmt.Errorf("invalid id: %w", err))
	}
```

We add context to the ID validation error and pass it to the `reply` adds the context information about the sending error. After that, we just need to return directly.

- Using the same structure (when creating the thing and sending the response despite success or failure):

```go
	token, err := i.thingProxy.Create(id, name, authorization)
	err = i.reply(id, token, err) // reply() will add the necessary context information and we try to send the response (success or failure).
	if err != nil {
                /* The response can't be sent and as the error already came with the sending error context information, 
                 we just need to add a bit of information about the whole procedure and return it. */
		return fmt.Errorf("error registering thing: %w", err)
	}
```

Reply:
Also, you have a good opportunity to refactor the `reply` method. Is "reply" a good name? Couldn't we move the part of it which is responsible for getting a pointer string from a received error to another function, something like `i.getNullableErrString(err)`?.

Reply:
Couldn't we change the `verifyThingID` to get rid of `ErrorIDLenght` and `ErrorIDInvalid` ? `errors.New()` is utterly  fine here.

Reply:
Couldn't we change the order of the functions? The majority of the code we're writing is placing the most important function at the beginning of the file. i.e. from the most important to the helpers, in a way that the code can be read as a text.

Reply:
@joaoaneto Excellent points! I've made some adjustments, waiting for review! :smile:

Reply:
Nice @esdrasjnr, the code improved a lot. Thanks!